app/controllers/katello/api/v2/systems_controller.rb

```diff
@@ -58,39 +59,29 @@ def organization_id_keys
     param_group :search, Api::V2::ApiController
     def index
+      respond(:collection => scoped_search(index_relation.uniq, :name, :asc))
+    end
+
+    def index_relation
       if params[:erratum_id] || params[:errata_ids]
         errata_ids = params.fetch(:errata_ids, [])
-        errata_ids << params[:erratum_id] if params[:erratum_id]
-        systems = systems_by_errata(errata_ids, params[:erratum_restrict_installable],
+        collection = systems_by_errata(errata_ids, params[:erratum_restrict_installable],
             params[:erratum_restrict_non_installable])
       else
-        systems = System.readable
+        collection = System.readable
       end
 
-      systems = systems.where(:content_view_id => params[:content_view_id]) if params[:content_view_id]
-      filters = [{:terms => {:uuid => systems.pluck("#{Katello::System.table_name}.uuid").compact}}]
+      collection = collection.where(:content_view_id => params[:content_view_id]) if params[:content_view_id]
+      collection = collection.where(:id => Organization.find(params[:organization_id]).systems.map(&:id)) if params[:organization_id]
       environment_ids = params[:organization_id] ? Organization.find(params[:organization_id]).kt_environments.pluck(:id) : []
-      environment_ids = environment_ids.empty? ? params[:environment_id] : environment_ids & [params[:environment_id].to_i] if params[:environment_id]
-      unless environment_ids.empty?
-        filters << {:terms => {:environment_id =>  environment_ids}}
-      end
-      if params[:host_collection_id]
-        filters << {:terms => {:host_collection_ids => [params[:host_collection_id]] }}
-      end
-      if params[:activation_key_id]
-        filters << {:terms => {:activation_key_ids => [params[:activation_key_id]] }}
-      end
-
-      filters << {:terms => {:uuid => System.all_by_pool_uuid(params['pool_id']) }} if params['pool_id']
-      filters << {:terms => {:uuid => [params['uuid']] }} if params['uuid']
-      filters << {:term => {:name => params['name'] }} if params['name']
-
-      options = {
-        :filters       => filters,
-        :load_records? => true,
-        :includes => [:content_view, :activation_keys, :environment]
-      }
-      respond_for_index(:collection => item_search(System, params, options))
+      environment_ids << params[:environment_id].to_i if params[:environment_id]
+      collection = collection.where(:id => KTEnvironment.where(:id => environment_ids).map(&:systems).flatten.map(&:id)) if environment_ids.any?
+      collection = collection.where(:id => HostCollection.find(params[:host_collection_id]).systems.map(&:id)) if params[:host_collection_id]
+      collection = collection.where(:id => ActivationKey.find(params[:activation_key_id]).systems.map(&:id)) if params[:activation_key_id]
```

Comment:
Similar comments to that above given systems has a AR relationship to AK and Host Collections.